plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go

```diff
@@ -0,0 +1,353 @@
+
+	// If username, password and clientId are provided, authenticate user before subscribing
+	// for data
+	if m.Username != "" && m.Password != "" && m.ClientID != "" {
+		lc := authentication.NewLoginClient(m.grpcClientConn)
+		loginReply, loginErr := lc.LoginCheck(context.Background(),
+			&authentication.LoginRequest{UserName: m.Username,
+				Password: m.Password, ClientId: m.ClientID})
+		if loginErr != nil {
+			return fmt.Errorf("E! Could not initiate login check: %v", err)
+		}
+
+		// Check if the user is authenticated. Bail if auth error
+		if !loginReply.Result {
+			return fmt.Errorf("E! Failed to authenticate the user")
+		}
+	}
+
+	c := telemetry.NewOpenConfigTelemetryClient(m.grpcClientConn)
+
+	for _, sensor := range m.Sensors {
+		wg.Add(1)
+		go func(sensor string, reportingRate uint32, acc telegraf.Accumulator) {
+			defer wg.Done()
+
+			spathSplit := strings.SplitN(sensor, " ", -1)
+			var slistStart int
+			var measurementName string
+			var pathlist []*telemetry.Path
+
+			// Extract measurement name and custom reporting rate if specified. Custom
+			// reporting rate will be specified at the beginning of sensor list,
+			// followed by measurement name like "1000ms interfaces /interfaces"
+			// where 1000ms is the custom reporting rate and interfaces is the
+			// measurement name. If 1000ms is not given, we use global reporting rate
+			// from sample_frequency. if measurement name is not given, we use first
+			// sensor name as the measurement name. If first or the word after custom
+			// reporting rate doesn't start with /, we treat it as measurement name
+			// and exclude it from list of sensors to subscribe
+			duration, err := time.ParseDuration(spathSplit[0])
+			if err == nil {
+				reportingRate = uint32(duration.Nanoseconds() / int64(time.Millisecond))
+				slistStart = 1
+			} else {
+				slistStart = 0
+			}
+
+			if len(spathSplit) <= slistStart {
+				acc.AddError(fmt.Errorf("E! No sensors are specified"))
+				return
+			}
+
+			// Word after custom reporting rate is treated as measurement name
+			measurementName = spathSplit[slistStart]
+
+			// If our word after custom reporting rate doesn't start with /, we treat
+			// it as measurement name. Else we treat it as sensor
+			if !strings.HasPrefix(measurementName, "/") {
+				slistStart += 1
+			}
+
+			if len(spathSplit) <= slistStart {
+				acc.AddError(fmt.Errorf("E! No valid sensors are specified"))
+				return
+			}
+
+			// List of sensors in this line
+			spathSplit = spathSplit[slistStart:]
+
+			// Iterate over our sensors and create pathlist to subscribe
+			for _, path := range spathSplit {
+				pathlist = append(pathlist, &telemetry.Path{Path: path,
+					SampleFrequency: reportingRate})
+			}
+
+			stream, err := c.TelemetrySubscribe(context.Background(),
+				&telemetry.SubscriptionRequest{PathList: pathlist})
+			if err != nil {
+				acc.AddError(fmt.Errorf("E! Could not subscribe: %v", err))
+				return
```

Comment:
Do we need to reconnect in the future, or can we assume if we couldn't subscribe we never will be able to?